l10n_it_fiscalcode/model/res_partner.py

```diff
@@ -1,31 +1,93 @@
 # -*- coding: utf-8 -*-
 # Copyright 2014 Associazione Odoo Italia (<http://www.odoo-italia.org>)
 # Copyright 2016 Andrea Gallina (Apulia Software)
+# Copyright 2016 Giuliano Lotta
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
-from odoo import models, fields, api
+from odoo import models, fields, api, _
+from odoo.exceptions import ValidationError
+
+import logging
+_logger = logging.getLogger(__name__)
+
+try:
+    from stdnum.it.codicefiscale import is_valid
+except ImportError:
+    _logger.warning(
+        'stdnum library not found. '
+        'If you plan to use it, please install the python-stdnum library '
+        'for Fiscal Code validation '
+        'from https://pypi.python.org/pypi/python-stdnum/')
 
 
 class ResPartner(models.Model):
+    """ Extends res.partner to add Italian Fiscal Code
+    """
+    # Private attributes
     _inherit = 'res.partner'
 
+    # Fields declaration
+    fiscalcode = fields.Char(
+        'Fiscal Code', size=16, help="Italian Fiscal Code")
+    is_soletrader = fields.Boolean(
+        string='Sole Trader',
+        default=False,
+        readonly=False,
+        help="Checked if company is a sole trader")
+
     @api.multi
-    def check_fiscalcode(self):
+    @api.onchange('is_company')
+    def onchange_iscompany(self):
+        """ if partner is switched from company to person,
+        is_soletrader is set to False because a simple private citizen
+        may not be a sole trader.
+        A proper warning is displayed.
+        """
         for partner in self:
-            if not partner.fiscalcode:
-                return True
-            elif len(partner.fiscalcode) != 16 and partner.individual:
-                return False
-            else:
-                return True
+            if partner.is_company is False:
```

Comment:
No, il test ha lo scopo di  verificare se is_company è impostato. Se compari is_company a False forzi il tipo di is_company a boolean. Il test che ti ho suggerito invece funziona con qualunque tipo contenga is_company.

Ad esempio prova questo

False if '' is False else True
False if '' else True

La stringa vuota non è boolean quindi il primo test fallisce, il secondo test invece (che poi è quello che ti ho suggerito) usa il concetto generico di python di Falso. Ora poichè il tuo scopo è solo verificare se la variabile è impostata meglio usare il concetto generico che rende il controllo funzionante indipendente dal tipo della variabile.

Fai attenzione che questo meccanismo di python ha anche un rovescio ed in tutti i casi in cui ti serve un controllo puntuale devi controllare per tipo.

ad esempio

a = []
if a .....

questo test fallisce ma "a" è impostata ad un oggetto, se lo scopo del test fosse verificare che a sia stato valorizzato allora il controllo corretto sarebbe if a is not None. Spero di aver chiarito se non è cosi' chiedi pure.